Display correct Trin version in CLI

akorchyn opened the pull request:
### What was wrong?
This PR resolves one issue and adds a small improvement.
The improvement is small; it replaces the ref Option with Option ref. It might seem unreasonable, but from the API standpoint, the ref option leaks implementation details; on the other hand, option ref just asks for optional data.

Resolves: [#9](https://github.com/ethereum/trin/issues/995)

### How was it fixed?
The shadow_rs collects data at compile time; I use this data to generate a `VERSION` const that is embedded in the clap library.
Everything is working on compile time.

### Example
```
yurtur@pc:~/projects/trin$ cargo run -- --version | tail -1 | sed "s/ /\//g"
    Finished dev [unoptimized + debuginfo] target(s) in 0.44s
     Running `target/debug/trin --version`
trin/0.2.2-e66c5c2c/linux-x86_64/rustc1.77.0
```

### To-Do
[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

KolbyML commented on ethportal-api/Cargo.toml:
```diff
@@ -50,6 +51,11 @@ tokio = { version = "1.14.0", features = ["full"] }
 ureq = { version = "2.5.0", features = ["json"] }
 url = "2.3.1"
 validator = { version = "0.13.0", features = ["derive"] }
+shadow-rs = "0.27"
+const_format = {version = "0.2.0", features = ["rust_1_64"]}
+
+[build-dependencies]
+shadow-rs = "0.27"
```
```suggestion
shadow-rs = "0.27"
const_format = {version = "0.2.0", features = ["rust_1_64"]}
```

Do we need shadow-rs in both build-dependcies and depends?

akorchyn replied:
Yeah, it collects the data in the build.rs and gives you access to it in the deps.

By the way, the const-format dep used indirectly (it is forwarded by the shadowrs, I declared it to get the split function that is under the feature gate)

KolbyML reviewed (commented):
Looks nice I will give the new imports a look in a few days and if it checks out merge it

ogenev reviewed (approved):
Thanks, LGTM 👍 !

ogenev commented:
In the future, we would like to report the version of the main Trin crate, not `ethportal-api` but this will require some more refactoring. I think this is good enough for this stage of development.

njgheorghita reviewed (commented):
It seems like there's an opportunity to consolidate all of our "versioning" utils. 
- [Here](https://github.com/ethereum/trin/blob/5e0658aa3365524a97e7830ec8ff1e67e4f1a4ca/trin-utils/build.rs#L11) we use the `GIT_HASH` env var to get the version
- and [here](https://github.com/ethereum/trin/blob/5e0658aa3365524a97e7830ec8ff1e67e4f1a4ca/trin-utils/src/version.rs#L5)

IIRC, this was required for compatibility with docker builds. (we set the `GIT_HASH` commit in the `docker build` command eg https://github.com/ethereum/trin/blob/5e0658aa3365524a97e7830ec8ff1e67e4f1a4ca/.circleci/config.yml#L69)

But, mostly just wanted to bring those to your attention, in case you had missed them. Maybe there's an opportunity here to use the same util in all of these scenarios? Ultimately, I'm not against using `shadow_rs` here, especially since it seems we'll need it for `build_os` and `rust_version`.

- Does `shadow_rs::build_info` work as expected within a docker container?
- Could we use `build_info` for all of the version utils (eg those inside `trin-utils`)?

akorchyn commented:
Hey, I'll look into the versioning files that you have mentioned. 
I'm new to the codebase, so I have missed them. I'll take a look to get them straightforward. 

In the meantime, I have fixed the nitpicks you mentioned.

Could you provide me with a helpful command to verify everything? It would be awesome.


By the way, keeping CI running for everyone who is not a contributor/maintainer is pretty insecure. I suggest you restrict the CI from outsiders, as I am until it's approved.

akorchyn commented:
Good catch with the docker. It actually doesn't work :c

akorchyn commented:
So, I have extracted shadow-rs into the utils crate.

Now, it should work with the env variable. Please note that I have used 12345678 as an env variable :)
![image](https://github.com/ethereum/trin/assets/42449190/3f73bbf0-ee2f-4826-83dd-765367568b4b)


I have added a hook to shadow rs to collect this particular env variable and added a const method to choose between:
* Collected from git
* Collected from env
* and `unknown` as recovery

@njgheorghita, I would love to get your thoughts on this. Just so you know, we will report the version coming from the utils crate this way. The proper way is to extract the CLI interface on the `trin` level. Otherwise, it's particularly unpleasant and hacky to collect proper versioning. If we moved the CLI interface to the trin level, we could get version data directly from the clap crate. However, I haven't looked at how coupled it is.

akorchyn commented:
Please let me know if I break the test. It looks like I'm not an issue, but still...
```
thread 'config::checkpoints::test::test_fetch_latest_checkpoints' panicked at light-client/src/config/checkpoints.rs:337:14:
called `Result::unwrap()` on an `Err` value: Failed to find max epoch from checkpoint slots
```

KolbyML commented:
Rebase the PR and that should be fixed

KolbyML commented:
@akorchyn if you could rebase the commit and rename the commit from `feat: Display` -> `feat: display` to fix the CI from erroring we can merge this

akorchyn commented:
@KolbyML done :P

KolbyML commented:
Thank you for your contributions
